Bug 1221303 - Firefox OS 2.5 landing page implementation

malena opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=1221303

malena commented:
Waiting on icons and urls for button and links.

alexgibson commented:
When viewing the older pages (e.g. `/firefox/os/2.0/`) I get an error `Exception Value: 'feed_links' is undefined`: http://dpaste.com/28MN3WH

alexgibson commented:
Visited link color needs a new value in the footer for better contrast

![footer-link-contrast](https://cloud.githubusercontent.com/assets/400117/11036678/ebf45006-86f1-11e5-8f7a-18430d0bbb78.png)

malena commented:
@alexgibson Hmm I'm not getting that error and I can preview those pages normally locally. This is also on demo5 and seems to be working there. (except for css updates that haven't been pushed)  https://www-demo5.allizom.org/b/en-US/firefox/os/2.0/

alexgibson commented:
Strange, I can view those pages ok on my master branch, but not this one for some reason?

alexgibson commented:
Form fields look a little odd being centered? Also the text below the submit button is a little hard to read.

<img width="675" alt="screen shot 2015-11-09 at 2 56 18 pm" src="https://cloud.githubusercontent.com/assets/400117/11036782/6a38ba92-86f2-11e5-8c59-cc11d45641c6.png">

malena commented:
@alexgibson Yeah I just caught the text align myself. Will push a few changes now.

alexgibson commented:
Worth putting these images through https://tinypng.com/ ?

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+    <h1>{{ _('Experience Firefox OS on your Android device')}}</h1>
+    <p>{{_('Introducing Firefox OS 2.5 Developer Preview, an experimental app that lets you to use Firefox OS on your Android device.')}}</p>
+    <div class="primary-cta-signup signup">
+      <a href="#TODO" class="cta-button" role="button">{{_('Get the Android App')}}</a>
```
Leaving `#TODO` here for reference

alexgibson replied:
Also add `rel="external"`?

malena replied:
Yeah that's the missing urls I still need. I'll add in the rel="external" since it's most likely the google play store.

alexgibson replied:
Also this probably shouldn't have `role="button"` assuming it's just a link?

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+  </div>
+  <div class="hero-image">
+    {% block phone_image %}
+      {{ high_res_img('firefox/os-new/phone.png') }}
```
We should probably include an empty alt attribute here if the image is just presentation?

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+{% endblock %}
+
+{% block features_copy %}
+  <p>{{_('Firefox OS is the first truly open mobile platform built entirely on Web technologies. If you can build for the Web using HTML and CSS, you can build Firefox OS.')}}
```
Missing closing `</p>`

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+      <div class="hub-container">
+        <h2>{{_('Visit the Participation Hub')}}</h2> 
+        <p>{{_('Want to make Firefox OS even better? Get all the resources, tools and documentation you need from the Participation Hub.')}}</p>
+        <p><a class="more" href="#TODO">{{_('Check it out')}}</a></p>
```
`#TODO` and also add `rel="external"` (assuming the link is external here too of course)

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+        <p><a class="more" href="#TODO">{{_('Check it out')}}</a></p>
+      </div>
+      <div class="hub-image">
+        {{ high_res_img('firefox/os-new/hub.png', {'width': '241', 'height': '261'}) }}
```
Add an empty `alt` attribute here too?

alexgibson replied:
This could probably also be a background image in CSS, but not a blocker :+1:

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+        <p><a class="more" href="#TODO">{{_('Visit MDN')}}</a></p>
+      </div>
+      <div class="mdn-image">
+        {{ high_res_img('firefox/os-new/mdn-logo.png', {'width': '301', 'height': '304'}) }}
```
Add an empty `alt` attribute here too? Likewise, this could probably also be a CSS background image maybe.

malena replied:
I would usually turn these into backgrounds, but left it the way template 2.0 was originally so it was less work updating. I was thinking these templates need a bit of clean up actually, the blocks are inconsistent, some wrap the whole section, some just the title and content or image. I can file a bug and do some refactoring to these later?

alexgibson replied:
Sure thing :+1:

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+      <div class="mdn-container">
+        <h2>{{_('Learn more at MDN')}}</h2>
+        <p>{{_('Want to do even more with Firefox OS? Head over to the Mozilla Developer Network to learn more about platform guides, Firefox OS add-ons and more.')}}</p>
+        <p><a class="more" href="#TODO">{{_('Visit MDN')}}</a></p>
```
`#TODO` and also add rel="external"

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+    <div class="content-wrapper">
+      <h2>{{_('Ready to try Firefox OS 2.5 Developer Preview on your Android device?')}}</h2>
+      <div class="primary-cta-signup signup">
+        <a href="#TODO" class="cta-button" role="button">{{_('Get the Android App')}}</a>
```
`#TODO`. Also this probably doesn't need `role="button"` if it's just a link.

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+      <h2>{{_('Ready to try Firefox OS 2.5 Developer Preview on your Android device?')}}</h2>
+      <div class="primary-cta-signup signup">
+        <a href="#TODO" class="cta-button" role="button">{{_('Get the Android App')}}</a>
+        <p>{{_('Want the full Firefox OS 2.5 Developer Preview experience?')}} <a href="#TODO">{{_('Flash your phone »')}}</a></p>
```
Normally we don't include `»` in the string itself. If you add a class of `more` to the link it should magically appear.

alexgibson commented on bedrock/firefox/templates/firefox/os/ver/2.5.html:
```diff
@@ -0,0 +1,109 @@
+{% block email_form %}
+  <section class="fxos-signup">
+    <div class="content-wrapper">
+      {{ email_newsletter_form(title="Sign up to get the latest Firefox OS news and more", submit_text="Sign up now") }}
```
Should these strings be hard coded in English?

malena replied:
No, I left it out because I was getting errors adding the l10n block inside. @alexgibson Wondering what the syntax for that would be?

alexgibson replied:
Try this:

      {% set newsletter_title = _('Send Firefox to your iOS device') %}
      {% set newsletter_submit = _('Sign up now') %}
      {{ email_newsletter_form(title=newsletter_title, submit_text=newsletter_submit) }}

alexgibson commented on media/css/firefox/os/firefox-os-new.less:
```diff
@@ -0,0 +1,491 @@
+
+    h2 {
+        .open-sans-light();
+        .font-size(3rem);
```
Nit: use `px` value for consistency?

alexgibson commented on media/css/firefox/os/firefox-os-new.less:
```diff
@@ -0,0 +1,491 @@
+                width: 140px;
+                height: 140px;
+                content: '';
+                background-image: url(/media/img/firefox/os-new/icons/features-sprite.png);
```
No need to include this if you're doing `.at2x` below. The mixin should generate both for you.

alexgibson commented on bedrock/firefox/templates/firefox/os/index.html:
```diff
@@ -207,9 +213,21 @@
+    {% block fxox_try %}
+    {% endblock %}
 
     {% set feed_links = firefox_os_feed_links(LANG) %}
```
I think this line and the conditional block below (`{% if feed_links %}`) need to move inside the `{% block fxox_news %}` block. Unless I do this, the other pages seem to throw the error mentioned earlier.

alexgibson commented:
Is it worth making a JS bundle for the 2.5 page? Seems like all that's needed is the family navigation for this page, and the rest is superfluous?

alexgibson commented:
Letter spacing on small text could do with a little more breathing room at small viewport?

![Uploading small-text.png…]()

alexgibson commented:
Could I suggest we rename `firefox-os-new.less` to something like `firefox-os-2-5.less`?

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -428,6 +428,17 @@
         ),
         'output_filename': 'css/firefox_os-bundle.css',
     },
+    'firefox_os_new': {
```
Could I suggest we rename this bundle to `firefox_os_2_5` or something a little more descriptive?

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -428,6 +428,17 @@
+            'css/libs/jquery.pageslide.css',
+            'css/firefox/family-nav.less',
+            'css/firefox/os/get_device.less',
+            'css/firefox/os/firefox-os-new.less',
```
We could also rename firefox-os-new.less to something like firefox-os-2-5.less?

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -428,6 +428,17 @@
+            'css/tabzilla/tabzilla-static.less',
+            'css/libs/jquery.pageslide.css',
+            'css/firefox/family-nav.less',
+            'css/firefox/os/get_device.less',
```
Also not sure this file is needed?

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -428,6 +428,17 @@
+        'source_filenames': (
+            'css/sandstone/sandstone-resp.less',
+            'css/tabzilla/tabzilla-static.less',
+            'css/libs/jquery.pageslide.css',
```
Is this one needed still too?

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -428,6 +428,17 @@
+            'css/firefox/os/get_device.less',
+            'css/firefox/os/firefox-os-new.less',
+        ),
+        'output_filename': 'css/firefox_os_new-bundle.css',
```
Bundle name could be `firefox_os_2_5-bundle.css`?

malena commented:
Yeah, I originally had named it firefox-os-2.5.less .. and then there were Travis CLI issues, and I thought it was the name.. so I changed it thinking that was the problem. I can definitely update name and bundles :D

alexgibson commented:
> Yeah, I originally had named it firefox-os-2.5.less

I'm guessing it might not have liked the extra `.`, but using something like a `-` should be ok.

alexgibson commented:
Would appreciate if @jgmize could take a quick look on the Python bits in this PR? (code was originally added by @pmclanahan in #3519)

pmclanahan commented:
Thanks Alex. And I took a look at the integration and don't see any issues. Would of course be good to get a second look. Nice work Malena :+1:

jgmize commented on bedrock/firefox/views.py:
```diff
@@ -337,11 +337,15 @@ def all_downloads(request, channel):
-        settings.FIREFOX_OS_COUNTRY_VERSIONS['default']
-    )
+    locale = l10n_utils.get_locale(request)
+    if locale == 'en-US':
```
please change this to 
```python
if locale == 'en-US' and waffle.switch('firefox-os-2.5'):
```
waffle is already imported near the top of this file, so that should be all we need, as discussed in the PR triage meeting.

pmclanahan replied:
it's actually `waffle.switch_is_active()`. There are examples elsewhere in this file if you need.

jgmize replied:
Oops, I was thinking of the template helper. thanks @pmclanahan, good catch.

jgmize commented on bedrock/firefox/tests/test_views.py:
```diff
@@ -313,11 +313,20 @@ def setUp(self):
+        request.locale = locale
         return views.firefox_os_geo_redirect(request)
 
+    def test_en_US_always_2_dot_5(self):
```
With the waffle switch, we'll need the following decorator above this line:
```python
@override_switch('firefox-os-2.5', active=True)
```

We'll also need to add `from waffle.testutils import override_switch` near the top of the file-- see http://waffle.readthedocs.org/en/stable/testing/automated.html for more details.

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,26 @@
+    'use strict';
+
+    $('.cta-button').on('click', function(e) {
+        e.preventDefault();
```
If we're adding a tracking pixel to this page, we should respect Do Not Track like we do elsewhere. We have a [handy global helper](https://github.com/mozilla/bedrock/blob/master/media/js/base/dnt-helper.js) that makes this easy in just a few lines of code. Something like:

    if (!window._dntEnabled()) {
        e.preventDefault();
        ftGoalTag14004();
    }

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,26 @@
+        var ftRand = Math.random() + "";
+        var num = ftRand * 1000000000000000000;
+        var ftGoalTagPix14004 = new Image();
+        ftGoalTagPix14004.src = "http://servedby.flashtalking.com/spot/2713;14004;1723/?spotName=submit_button& cachebuster="+num;
```
Let's use `https://` here

alexgibson replied:
Also let's use single quotes on strings throughout this file for consistency?

alexgibson replied:
There also seems to be a stray space in this string before `cachebuster` that needs removing.

Adding a space either side of `+` for readability would be nice too.

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,26 @@
+        var num = ftRand * 1000000000000000000;
+        var ftGoalTagPix14004 = new Image();
+        ftGoalTagPix14004.src = "http://servedby.flashtalking.com/spot/2713;14004;1723/?spotName=submit_button& cachebuster="+num;
+        ftGoalTagPix14004.onload = ftLoaded14004;
```
This is currently breaking the button for any user with Tracking Protection enabled, since it prevents the image from loading and the onload event doesn't fire. (oh, the irony). I suggest we just use a short delay using `setTimeout` instead.

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,26 @@
+
+    function ftLoaded14004() { 
+        // #TODO change this link
+        top.location.href="http://www.google.com"; 
```
Nit: space either side of `+` and single quotes for consistency.

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,26 @@
+    });
+
+    function ftGoalTag14004(){
+        var ftRand = Math.random() + "";
```
Single quotes

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,31 @@
+        
+        ftGoalTagPix14004.src = 'https://servedby.flashtalking.com/spot/2713;14004;1723/?spotName=submit_button&cachebuster=' + num;
+
+        window.setTimeout(function(){
```
You should be able to change this to just `setTimeout(ftLoaded14004, 200);`

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    function ftLoaded14004() { 
+        top.location.href = 'https://d2yw7jilxa8093.cloudfront.net/B2GDroid-mozilla-central-nightly-latest.apk'; 
```
Since we're not using multiple windows/iframes here, you can probably just change this to `window.location.href`

alexgibson commented on media/js/firefox/os/firefox-os-2.5.js:
```diff
@@ -0,0 +1,29 @@
+
+        ftGoalTagPix14004.src = 'https://servedby.flashtalking.com/spot/2713;14004;1723/?spotName=submit_button&cachebuster=' + num;
+
+        setTimeout(ftLoaded14004, 200);
```
Given some testing it looks like 200ms might not be enough - changing it to 300 seems to be about the right amount here?

alexgibson commented:
As it stands, this gets an r+wc from me :bear:  (just need to adjust the timing a little as commented)

malena commented:
@alexgibson Got the final graphics (two circles). Hero isn't changing. We will most likely be filing a bug to update it later. So hopefully this is the last update!!

alexgibson commented:
r+ nice work! Care to squarsh? :dolphin:

alexgibson commented:
Ah, just spotted one more CSS issue - the thank you message for the newsletter is very large?

![thankyou](https://cloud.githubusercontent.com/assets/400117/11067916/cbbf5756-87c6-11e5-949d-d01f8229f25b.png)

malena commented:
@alexgibson I've reduced the font size, tested on samsung s4. Not sure how you get that rounded corner white background text overflowing in your screenshot?

alexgibson commented:
@malena sorry I posted the wrong screenshot initially (too much code review). I updated it here to the correct one

malena commented:
@alexgibson Ah okay, yeah the font size should be reasonably smaller now.

alexgibson commented:
r+ :racehorse:

malena commented:
Addressed comment in bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1221303#c13
Remove "Developer Preview" from the "Want the full Firefox OS 2.5 Developer Preview experience? Flash your phone" link at the bottom of the page.
